File: src/main/java/com/github/njuro/jard/ban/BanRestController.java
package com.github.njuro.jard.ban;

import com.github.njuro.jard.ban.dto.BanDto;
import com.github.njuro.jard.ban.dto.BanForm;
import com.github.njuro.jard.common.Mappings;
import com.github.njuro.jard.config.security.methods.HasAuthorities;
import com.github.njuro.jard.user.UserAuthority;
import java.util.List;
import javax.servlet.http.HttpServletRequest;
import javax.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping(Mappings.API_ROOT_BANS)
@SuppressWarnings("MVCPathVariableInspection")
public class BanRestController {

  private final BanFacade banFacade;

  @Autowired
  public BanRestController(BanFacade banFacade) {
    this.banFacade = banFacade;
  }

  @PostMapping
  @HasAuthorities({UserAuthority.MANAGE_BANS, UserAuthority.VIEW_IP})
  public ResponseEntity<BanDto> createBan(@RequestBody @Valid BanForm banForm) {
    return ResponseEntity.status(HttpStatus.CREATED).body(banFacade.createBan(banForm));
  }

  @GetMapping("/me")
  public BanDto getOwnBan(HttpServletRequest request) {
    return banFacade.getActiveBan(request.getRemoteAddr());
  }

  @GetMapping
  @HasAuthorities({UserAuthority.MANAGE_BANS, UserAuthority.VIEW_IP})
  public List<BanDto> getAllBans() {
    return banFacade.getAllBans();
  }

  @PutMapping(Mappings.PATH_VARIABLE_BAN)
  @HasAuthorities({UserAuthority.MANAGE_BANS, UserAuthority.VIEW_IP})
  public BanDto editBan(BanDto oldBan, @RequestBody @Valid BanForm banForm) {
    return banFacade.editBan(oldBan, banForm);
  }

  @PutMapping(Mappings.PATH_VARIABLE_BAN + "/unban")
  @HasAuthorities({UserAuthority.MANAGE_BANS, UserAuthority.VIEW_IP})
  public BanDto unban(BanDto ban, @RequestBody @Valid UnbanForm unbanForm) {
    return banFacade.unban(ban, unbanForm);
  }
}

File: src/main/kotlin/com/github/njuro/jard/rewrite/ban/BanRestController.kt
package com.github.njuro.jard.rewrite.ban

import com.github.njuro.jard.rewrite.ban.dto.BanDto
import com.github.njuro.jard.rewrite.ban.dto.BanForm
import com.github.njuro.jard.rewrite.ban.dto.UnbanForm
import com.github.njuro.jard.rewrite.common.API_ROOT_BANS
import com.github.njuro.jard.rewrite.common.PATH_VARIABLE_BAN
import com.github.njuro.jard.rewrite.config.security.methods.HasAuthorities
import com.github.njuro.jard.rewrite.user.UserAuthority.MANAGE_BANS
import com.github.njuro.jard.rewrite.user.UserAuthority.VIEW_IP
import org.springframework.http.HttpStatus.CREATED
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import javax.servlet.http.HttpServletRequest
import javax.validation.Valid

@RestController
@RequestMapping(API_ROOT_BANS)
class BanRestController(private val banFacade: BanFacade) {
    @PostMapping
    @HasAuthorities(MANAGE_BANS, VIEW_IP)
    fun createBan(@RequestBody banForm: @Valid BanForm): ResponseEntity<BanDto> =
        ResponseEntity.status(CREATED).body(banFacade.createBan(banForm))

    @GetMapping("/me")
    fun getOwnBan(request: HttpServletRequest): BanDto {
        return banFacade.getActiveBan(request.remoteAddr)
    }

    @HasAuthorities(MANAGE_BANS, VIEW_IP)
    @GetMapping
    fun getAllBans(): List<BanDto> = banFacade.getAllBans()

    @PutMapping(PATH_VARIABLE_BAN)
    @HasAuthorities(MANAGE_BANS, VIEW_IP)
    fun editBan(oldBan: BanDto, @RequestBody banForm: @Valid BanForm): BanDto =
        banFacade.editBan(oldBan, banForm)


    @PutMapping("$PATH_VARIABLE_BAN/unban")
    @HasAuthorities(MANAGE_BANS, VIEW_IP)
    fun unban(ban: BanDto, @RequestBody unbanForm: @Valid UnbanForm): BanDto =
         banFacade.unban(ban, unbanForm)

}
